### DISCUSSION TITLE
Subscription httpapi

### DISCUSSION THREAD

**Discussion by kalwinskidawid** [avantrec/soco-cli#33] [Q&A] [answered]
Hey, I have three questions: 
Is it possible 
 - get summarized information about the current state of the device (i.e. is it playing, current song, next song)?
- subscribe to information (real time info) with summary (is it playing, current song, next song) via http/weboscket?
- download the album_art of the next track (I would like to have the current album_art of the current and next track)?

**Comment by pwt** [comment]
In general, there's lots of information on the currently playing track, but not on the next track -- at least not via Sonos-native methods. However, if Sonos is playing from the queue, it's possible to use the `queue` action to get information about upcoming tracks.

> * get summarized information about the current state of the device (i.e. is it playing, current song, next song)?

Take a look at the `track` action.

> * subscribe to information (real time info) with summary (is it playing, current song, next song) via http/weboscket?

Take a look at the `track_follow` and `track_follow_compact` actions. There's no HTTP or websocket support.

> * download the album_art of the next track (I would like to have the current album_art of the current and next track)?

You can get a URL to the current track's album art using the `album_art` action, but not for the next track.

**Comment by kalwinskidawid** [comment]
Thanks @pwt 
In the future will it be possible to use track_follow via http? 
I asked about next track album_art because I saw sth like that in FHEM:
![image](https://user-images.githubusercontent.com/9535316/162423047-6ccf8096-ad71-4581-aa38-4c2652d154e7.png)

**Reply by pwt** [reply]
- I'm not planning to add `track_follow` to the HTTP API.
- Perhaps it's possible to use the Sonos API to get more information on the next track ... I'll take a look when I get some time.

**Comment by kalwinskidawid** [comment]
Actually one more question, is it possible to secure the api? For example by giving key in GET header

**Reply by pwt** [reply]
Not at the moment, and it's not something I'm planning. I figured that there's not much point, as the Sonos UPnP API is completely unsecured.

I guess if one is making the HTTP API accessible remotely then it would be useful, but then I'd want TLS as well as password security.

**Reply by kalwinskidawid** [reply]
You're right, the UPnP API is completely insecure. Yea, now I realized that was stupid question :) 
I'm still trying to figure out this api, because I would like to connect it to my java application and it would be useful to receive states in real time. I was thinking about httplongpoling, but then if I have X speakers I would have to compile X instances of httpoling, is there any address where all devices are visible with current state(playing, track info)?

**Reply by pwt** [reply]
The `sysinfo` action will give you the playback state for all speakers as part of its output, but not the track data.

The `wait_end_track` action will wait until a speaker's current track ends before returning. This can be used instead of polling.

Since you're working in Java it might be easier just to use a Java library for the Sonos UPnP API. There don't seem to be any that are under active development, but you might be able to leverage/adapt enough code for your use case from the couple that do exist.

**Reply by kalwinskidawid** [reply]
Are you thinking about KilianB/Java-Sonos-Controller? I've been checking it out and was going to use it, but I'm having trouble with two old library. 
Is it possible to use `wait_end_track `via httpapi? Could I ask for some example?

**Reply by pwt** [reply]
> Are you thinking about KilianB/Java-Sonos-Controller? I've been checking it out and was going to use it, but I'm having trouble with two old library.

Or this one: https://github.com/nightowlengineer/sonos-api-java

> Is it possible to use `wait_end_track` via httpapi? Could I ask for some example?

Yes, e.g.: `http://<ip_address>:8000/<speaker_name>/wait_end_track` won't return until a currently playing track ends. You can also create macros in which `wait_end_track` is one of the commands.

Why don't you experiment a bit? SoCo-CLI is easy to set up and use, and is very well documented.

**Reply by kalwinskidawid** [reply]
> Or this one: https://github.com/nightowlengineer/sonos-api-java

I looked at it, unfortunately the API from Sonos is pointless to me because everything goes through their cloud. It seems more reliable for me to send SOAP messages over the local network.

> Why don't you experiment a bit? SoCo-CLI is easy to set up and use, and is very well documented.

The documentation is great! Thanks for this! 
I was testing `wait_end_track `and just wondered why I wasn't getting a response :D

**Comment by kalwinskidawid** [comment]
Hey, one more question :) 
How do you recognize what kind of song is currently playing (tunein, spotify etc) ? 
I don't want to take your code apart, which is why this question.

Thank you for your help so far!
PS. I moved on Java library, however your project is still helpful.

**Reply by pwt** [reply]
Unfortunately, I don't know. It's probably somewhere in the events received from `avTransport` subscriptions, but I don't capture it.

The implementation of the `track` action in SoCo-CLI may help:
  https://github.com/avantrec/soco-cli/blob/f3a5bb82b59aaf3438cf9663e6894abed120481c/soco_cli/action_processor.py#L368

**Comment by pwt** [comment, marked as answer]
Closing as 'answered'.
